### PR TITLE
Add `--timing` flag, phase timing to @dynamo_timed

### PR DESCRIPTION
**Pull request by voznesenskym**
Ex output:
```
 TIMING:
 entire_frame_compile:8.574629999999999
 backend_compile:5.26806
```

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -86,20 +85,81 @@ def profile_wrapper(*args, **kwargs):
...
+# Total time: 25.00655s
+def print_time_report():
+    total = 0
+    print("TIMING:")
+    for frame, timings in frame_phase_timing.items():
+        print(f"  frame #{frame}:")
+        for key, time in timings.items():
+            total += time
+            print(f"    {key} : {time}")
+    print(f"Total time: {total}s")
```
Comment (voznesenskym) [inline torch/_dynamo/utils.py:121]: this is actually sus cause of nesting

File: benchmarks/dynamo/final.csv [hunk 2]
```diff
@@ -0,0 +1 @@
+,8383b5c488399f2ae295c7c0f993bdd353dfd75c,,,,,
```
Comment (ezyang) [inline benchmarks/dynamo/final.csv:1]: delete (and add .csv to gitignore if you need)

File: torch/_dynamo/utils.py [hunk 3]
```diff
@@ -86,20 +86,86 @@ def profile_wrapper(*args, **kwargs):
...
+        @wraps(func)
+        def time_wrapper(*args, **kwargs):
+            key = func.__qualname__
+            if key not in compilation_metrics:
+                compilation_metrics[key] = []
+            t0 = time.time()
+            r = func(*args, **kwargs)
+            time_spent = time.time() - t0
+            # print(f"Dynamo timer: key={key}, latency={latency:.2f} sec")
+            compilation_metrics[key].append(frame_phase_timing)
```
Comment (ezyang) [inline torch/_dynamo/utils.py:153]: bug

File: torch/_dynamo/utils.py [hunk 4]
```diff
@@ -86,20 +86,86 @@ def profile_wrapper(*args, **kwargs):
...
+#
+# Would show up as an entry in our timing dict:
+# OrderedDict([('bar.<locals>._foo', [0.083690, 0.23949, 3.1425e-05])])
+# This is extremely useful for granular debugging.
+#
+# For a higher-level mode, pass a phase_name into dynamo_timed
+# phase_names record an extra record into a separate compilation timing structure,
+# one keyed on frame+name rather than function.
+# The frame is incremented outside of this function, in def increment_frame() above.
+def dynamo_timed(original_function=None, phase_name=None):
```
Comment (ezyang) [inline torch/_dynamo/utils.py:142]: can we do phase name with an enum or something

File: torch/_dynamo/utils.py [hunk 5]
```diff
@@ -86,20 +86,86 @@ def profile_wrapper(*args, **kwargs):
...
+            # print(f"Dynamo timer: key={key}, latency={latency:.2f} sec")
+            compilation_metrics[key].append(frame_phase_timing)
+            if phase_name:
+                frame_key = str(curr_frame)
+                if frame_key not in frame_phase_timing:
+                    frame_phase_timing[frame_key] = {}
+                assert (
+                    phase_name not in frame_phase_timing[frame_key]
+                ), f"Duplicate phase name {phase_name} for frame {frame_key}"
+                frame_phase_timing[frame_key][phase_name] = round(time_spent, 5)
```
Comment (ezyang) [inline torch/_dynamo/utils.py:161]: why would you round 🤔

File: torch/_dynamo/utils.py [hunk 6]
```diff
@@ -86,20 +86,86 @@ def profile_wrapper(*args, **kwargs):
...
+    print("TIMING:")
+    for frame, timings in frame_phase_timing.items():
+        for key, timing in timings.items():
+            total += timing
+            if key not in total_by_key:
+                total_by_key[key] = timing
+            else:
+                total_by_key[key] += timing
+    for key, value in total_by_key.items():
+        print(f"{key}:{round(value, 5)}")
```
Comment (ezyang) [inline torch/_dynamo/utils.py:123]: My preference is to have this as a single line, to make it easier to grep for by automated tools

**Review by ezyang — APPROVED** [review]
seems fine I guess

File: torch/_dynamo/utils.py (hunk 4 above)
Comment (voznesenskym) [reply torch/_dynamo/utils.py:142]: mmmm I plan to add many as I debug, I'd rather have this be more fluid...

File: torch/_dynamo/utils.py (hunk 6 above)
Comment (voznesenskym) [reply torch/_dynamo/utils.py:123]: sure

**Comment by malfet** [comment]
> @pytorchbot merge -f "local lintrunner -a doesn't show a diff. lintrunner init run, still no dice. I give up. yolo"

Not sure what was the point of doing this? If commit breaks CI it will be reverted, it's just the matter of time...

**Comment by albanD** [comment]
btw some lints (like flake8) cannot be automatically fixed. So it is expected that `-a` will not fix everything in some cases!

**Comment by voznesenskym** [comment]
> btw some lints (like flake8) cannot be automatically fixed. So it is expected that `-a` will not fix everything in some cases!

I was plagued with two problems:

1) lintrunner -a does not look at everything (need lintrunner -m master) 
2) Unrelated failures - In this case, the failures are on lines I did not touch. @malfet advised that it is a bad base. (current state)

**Comment by voznesenskym** [comment]
@pytorchbot merge -m "Linter failure for code I did not touch"

image

**Comment by voznesenskym** [comment]
@pytorchbot merge -f "Linter failure for code I did not touch"

image

**Comment by kit1980** [comment]
> "Linter failure for code I did not touch"

Looks like the MyPy failure is because the the decorator `dynamo_timed` was removed from `_convert_frame_assert`, changing its type.

**Comment by voznesenskym** [comment]
Oof sorry about this, I've been banging into the linter like an idiot all week

**Pull request closed without merging**